b0c598742fe9b8ffe5fcdfdc9adce34708907379 breaks build on OS X, does not eliminate warnings

Trying to build current master on OS X 10.11.6 with Xcode 8 yields the following early on in the build process:

```
../lex.c:137:19: error: use of undeclared identifier 'deftbl'
        if ((tp = lookup(deftbl, token, NULL)) != NULL) {
                         ^
../lex.c:144:24: error: use of undeclared identifier 'keytbl'
        else if ((tp = lookup(keytbl, token, NULL)) == NULL) {
                              ^
```

git-blaming reveals that commit b0c598742fe9b8ffe5fcdfdc9adce34708907379 removed the declarations of `deftbl` and `keytbl` from `yylex()` in eqn. If I manually add them back, we get a bit more progress:

```
../lookup.c:223:1: error: conflicting types for 'lookup'
lookup(tbl **tblp, const char *name, const char *defn)  /* find name in ...
^
../e.h:150:6: note: previous declaration is here
tbl *lookup(tbl **, char *, char *);
     ^
```

Readding the `const` qualifiers to the parameters in `lookup.c` brings us even further:

```
../paren.c:166:1: error: conflicting types for 'brack'
brack(int m, const char *t, const char *c, const char *b) {
^
../e.h:164:6: note: previous declaration is here
void brack(int, char *, char *, char *);
     ^
```

but at this point I stopped and did the git blame, and noticed that the commit in question was trying to fix constness warnings, which still show up for me among the other noise.

Reverting the commit entirely causes eqn to build without any warnings or errors. It then breaks on dpost with similar errors, from a similar commit.

Example make output for one file's compilation, to show compiler flags:

```
cc   -D_GNU_SOURCE -I. -I.. -I../../include -DHAVE_STRLCPY -DHAVE_STRLCAT -DHAVE_WCSLCPY -c ../lex.c
```

clang version:

```
Apple LLVM version 8.0.0 (clang-800.0.38)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

cc @bapt

--- a/eqn/e.h
+++ b/eqn/e.h
@@ -147,7 +147,7 @@ char *strsave(char *);
 void include(void);
 void delim(void);
 /* lookup.c */
-tbl *lookup(tbl **, char *, char *);
+tbl *lookup(tbl **, const char *, const char *);
 void init_tbl(void);
 /* mark.c */
 void mark(int);
@@ -161,7 +161,7 @@ void move(int, int, int);
 void boverb(int, int);
 /* paren.c */
 void paren(int, int, int);
-void brack(int, char *, char *, char *);
+void brack(int, const char *, const char *, const char *);
 /* pile.c */
 void lpile(int, int, int);
 /* shift.c */

--- a/eqn/e.h
+++ b/eqn/e.h
@@ -92,8 +92,8 @@ extern int	markline;	/* 1 if this EQ/EN contains mark or lineup */
 extern char	*progname;
 
 typedef struct s_tbl {
-	char	*name;
-	char	*defn;
+	const char	*name;
+	const char	*defn;
 	struct s_tbl *next;
 } tbl;
 extern  char    *spaceval;  /* use in place of normal \x (for pic) */
@@ -137,7 +137,7 @@ void error(int, const char *, ...);
 /* lex.c */
 int gtc(void);
 int openinfile(void);
-void pbstr(register char *);
+void pbstr(register const char *);
 int yylex(void);
 int getstr(char *, register int);
 int cstr(char *, int, int);

--- a/eqn/lex.c
+++ b/eqn/lex.c
@@ -94,6 +94,7 @@ int
 yylex(void) {
 	register int c;
 	tbl *tp;
+	extern tbl *keytbl[], *deftbl[];
 
   beg:
 	while ((c=gtc())==' ' || c=='\n')

--- a/eqn/lex.c
+++ b/eqn/lex.c
@@ -77,9 +77,9 @@ openinfile(void)
 }
 
 void
-pbstr(register char *str)
+pbstr(register const char *str)
 {
-	register char *p;
+	register const char *p;
 
 	p = str;
 	while (*p++);

--- a/eqn/text.c
+++ b/eqn/text.c
@@ -42,6 +42,7 @@ text(int t,char *p1) {
 	int c;
 	const char *p;
 	tbl *tp;
+	extern tbl *restbl;
 
 	yyval.token = oalloc();
 	ebase[yyval.token] = 0;

--- a/pic/for.c
+++ b/pic/for.c
@@ -28,7 +28,7 @@ typedef struct {
 static For	forstk[10];	/* stack of for loops */
 static For	*forp = forstk;	/* pointer to current top */
 
-void	setfval(char *, double);
+void	setfval(const char *, double);
 void	nextfor(void);
 
 void forloop(char *var, double from, double to, int op,

--- a/pic/pic.h
+++ b/pic/pic.h
@@ -187,7 +187,7 @@ void	copyfile(char *);
 void	copydef(struct symtab *);
 void	definition(const char *);
 void	freesymtab(struct symtab *);
-void	setfval(char *s, double f);
+void	setfval(const char *s, double f);
 struct symtab *copythru(const char *);
 #ifdef	FLEX_SCANNER
 int	xxinput(void);

--- a/pic/symtab.c
+++ b/pic/symtab.c
@@ -41,7 +41,7 @@ double getfval(const char *s)	/* return float value of variable s */
 	return y.f;
 }
 
-void setfval(char *s, double f)	/* set variable s to f */
+void setfval(const char *s, double f)	/* set variable s to f */
 {
 	struct symtab *p;
 

--- a/refer/refer..c
+++ b/refer/refer..c
@@ -164,7 +164,7 @@ char *trimnl(char *);
 /* refer5.c */
 void putsig(int, char **, int, char *, char *, int);
 char *fpar(int, char **, char *, size_t, int, int, int);
-void putkey(int, char **, int, char *);
+void putkey(int, char **, int, const char *);
 void tokeytab(const char *, int);
 int keylet(char *, int);
 void mycpy(char *, const char *);

--- a/refer/refer5.c
+++ b/refer/refer5.c
@@ -240,7 +240,7 @@ fpar (int nf, char **flds, char *out, size_t outsiz __unused, int c, int seq,
 }
 
 void
-putkey(int nf, char **flds, int nref, char *_keystr)
+putkey(int nf, char **flds, int nref, const char *_keystr)
 {
 	char t1[50], *sf;
 	int ctype, i, count;

--- a/troff/n3.c
+++ b/troff/n3.c
@@ -96,7 +96,7 @@ static void	caseindex(void);
 static void	caseasciify(void);
 static void	caseunformat(int);
 static int	getls(int, int *, int);
-static void	addcon(int, char *, void(*)(int));
+static void	addcon(int, const char *, void(*)(int));
 
 static const struct {
 	const char	*n;
@@ -2353,7 +2353,7 @@ makerq(const char *name)
 }
 
 static void
-addcon(int _t, char *rs, void(*f)(int))
+addcon(int _t, const char *rs, void(*f)(int))
 {
 	int	n = hadn;
 

--- a/troff/troff.d/afm.c
+++ b/troff/troff.d/afm.c
@@ -992,8 +992,8 @@ afmget(struct afmtab *a, char *contents, size_t size)
  * troff and dpost need it in combination with AFM support.
  */
 void
-makefont(int nf, char *devfontab, char *devkerntab, char *devcodetab,
-		char *devfitab, int nw)
+makefont(int nf, const char *devfontab, const char *devkerntab,
+    const char *devcodetab, const char *devfitab, int nw)
 {
 	int	i;
 

--- a/troff/troff.d/afm.h
+++ b/troff/troff.d/afm.h
@@ -123,7 +123,7 @@ extern	int	afmget(struct afmtab *, char *, size_t);
 extern	int	otfget(struct afmtab *, char *, size_t);
 extern	struct namecache	*afmnamelook(struct afmtab *, const char *);
 extern	int	afmgetkern(struct afmtab *, int, int);
-extern	void	makefont(int, char *, char *, char *, char *, int);
+extern	void	makefont(int, const char *, const char *, const char *, const char *, int);
 extern	int	unitconv(int);
 extern	void	afmalloc(struct afmtab *, int);
 extern	void	afmremap(struct afmtab *);

--- a/troff/troff.d/dpost.d/dpost.c
+++ b/troff/troff.d/dpost.d/dpost.c
@@ -1926,7 +1926,7 @@ loadfont (
     char	*fpout = NULL;		/* for reading *s file */
     int		fin;			/* for reading *s.afm file */
     int		nw;			/* number of width table entries */
-    char	*p;
+    const char	*p;
     char	*path;
     size_t	l;
 
@@ -2167,7 +2167,7 @@ fontprint (
 
 const char *
 mapfont (
-    char *name			/* troff wanted this font */
+    const char *name			/* troff wanted this font */
 )
 
 

--- a/troff/troff.d/dpost.d/gen.h
+++ b/troff/troff.d/dpost.d/gen.h
@@ -98,7 +98,7 @@ void loadfont(int, const char *, char *, int, int);
 void loadspecial(void);
 void loaddefault(void);
 void fontprint(int);
-const char *mapfont(char *);
+const char *mapfont(const char *);
 void getdevmap(void);
 char *mapdevfont(char *);
 void reset(void);

--- a/troff/troff.d/dpost.d/gen.h
+++ b/troff/troff.d/dpost.d/gen.h
@@ -94,7 +94,7 @@ void account(void);
 void conv(register FILE *);
 void devcntrl(FILE *);
 void fontinit(void);
-void loadfont(int, char *, char *, int, int);
+void loadfont(int, const char *, char *, int, int);
 void loadspecial(void);
 void loaddefault(void);
 void fontprint(int);

--- a/troff/troff.d/dpost.d/request.c
+++ b/troff/troff.d/dpost.d/request.c
@@ -138,7 +138,7 @@ dumprequest(
 
 
     char	*want,			/* look for this string */
-    char	*file,			/* in this file */
+    const char	*file,			/* in this file */
     FILE	*fp_out			/* and write the value out here */
 )
 

--- a/troff/troff.d/fontmap.c
+++ b/troff/troff.d/fontmap.c
@@ -94,8 +94,8 @@ rdftmap(char *path) {
 	free(buf);
 }
 
-char *
-mapft(char *name) {
+const char *
+mapft(const char *name) {
 	struct bst_node *n;
 	if (map.root && !bst_srch(&map, (union bst_val)(void *)name, &n))
 		name = n->data.p;

--- a/troff/troff.d/fontmap.h
+++ b/troff/troff.d/fontmap.h
@@ -1,2 +1,2 @@
 void rdftmap(char *);
-char *mapft(char *);
+const char *mapft(const char *);

--- a/troff/troff.d/pt.h
+++ b/troff/troff.d/pt.h
@@ -85,7 +85,7 @@ extern void	ptout(register tchar);
 extern tchar	*ptout0(tchar *, tchar *);
 extern void	ptps(void);
 extern void	ptfont(void);
-extern void	ptfpcmd(int, char *, char *, int);
+extern void	ptfpcmd(int, const char *, char *, int);
 extern void	ptlead(void);
 extern void	ptesc(void);
 extern void	newpage(int);

--- a/troff/troff.d/t10.c
+++ b/troff/troff.d/t10.c
@@ -824,7 +824,7 @@ ptfont(void)
 }
 
 void
-ptfpcmd(int f, char *s, char *path, int flags)
+ptfpcmd(int f, const char *s, char *path, int flags)
 {
 	if (ascii)
 		return;

--- a/troff/troff.d/t6.c
+++ b/troff/troff.d/t6.c
@@ -843,7 +843,8 @@ findft(register int i, int required)
 {
 	register int k;
 	int nk;
-	char	*mn, *mp;
+	const char	*mn;
+	char *mp;
 
 	if ((k = i - '0') >= 0 && k <= nfonts && k < smnt && fontbase[k])
 		return(k);
@@ -1666,7 +1667,8 @@ casefps(void)
 int
 setfp(int pos, int f, char *truename)	/* mount font f at position pos[0...nfonts] */
 {
-	char longname[4096], *shortname, *ap;
+	char longname[4096], *ap;
+	const char *shortname;
 	char *fpout;
 	int i, nw;
 

